### PR TITLE
Partially shared volumes hide data from databrary community - backend (rebased)

### PR DESCRIPTION
**Pull request by kanishka-azimi**
DO NOT MERGE . (Needs light testing). Aiming to merge on Friday or next week.

These changes finish up an earlier sweeping change to volume sharing levels. The original change modified volume sharing levels from

private
public

to

private
partially shared
public

When partially shared, all volume data besides overview information is hidden to public viewers as well as databrary community member viewers. The original implementation only hid data from public viewers. These changes implement hiding from databrary community member viewers.

The core of this change can be observed through the (now poorly named) volumeIsPublicRestricted function. It observes the computed permission for the current user "volumeRolePolicy" and decides whether data should be hidden or not. VolumeRolePolicy consists of a role ("public", "shared", "readers", "editors", ...) and a refinement of the role with a policy. For most roles, there are no policies defined. For the "public viewer" and "shared viewer" roles, either the "restricted policy" or no policy can be specified.

The computed role/policy value for the current user on a given volume comes from multiple sources. The sources can be summarized as "direct access", "direct access via lab (authorized investigator of lab)", and "general group access". The role and policy for the current viewer is computed by taking the highest role/policy the user can achieve through the different sources. One is only allowed to annotate a role with a restricted policy when the role is being assigned through the "general group access" source.

A partially shared volume means the volume has been shared with the "public" group in the role "public" with policy "restricted", as well as being shared with the "databrary community" group in the role "shared" with policy "restricted".

### CODE REVIEW

File: src/Databrary/Controller/VolumeAccess.hs [hunk 1]
```diff
@@ -58,10 +58,13 @@ postVolumeAccess = action POST (pathJSON >/> pathId </> pathVolumeAccessTarget)
...
-      if ap == ((partyId . partyRow) nobodyParty) && individual == PermissionPUBLIC
+      if (ap, individual) `elem` [(getPartyId nobodyParty, PermissionPUBLIC), (getPartyId rootParty, PermissionSHARED)]
       then do
-        _ <- "share_full" .:> (deformCheck "Required" (not . (== FormDatumNone)) =<< deform) -- convulated way of requiring
-        Just <$> ("share_full" .:> deform)
+        if individual == PermissionSHARED -- TODO: remove me, temporary until front end updated
+        then pure (Just False)
+        else do
+          _ <- "share_full" .:> (deformCheck "Required" (not . (== FormDatumNone)) =<< deform) -- convulated way of requiring
+          Just <$> ("share_full" .:> deform)
```
Comment (kanishka-azimi) [inline src/Databrary/Controller/VolumeAccess.hs]: Deal with TODO as soon as front end is modified.

File: schema/20180523-access-sharefull-contd.sql [hunk 2]
```diff
@@ -0,0 +1,69 @@
+ALTER TABLE "volume_access" DROP CONSTRAINT "volume_access_check_full1";
+ALTER TABLE "volume_access" DROP CONSTRAINT "volume_access_check_full2";
+ALTER TABLE "volume_access" ADD CONSTRAINT "volume_access_check_full3"
+  CHECK ( party = -1 OR party = 0 OR share_full is null );
+-- expanding full2 for party 0 would be too complicated
+
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: First real change below.

File: src/Databrary/Controller/Volume.hs [hunk 3]
```diff
@@ -135,8 +135,9 @@ leftJoin p (a:al) b = uncurry (:) $ (,) a *** leftJoin p al $ span (p a) b
 
 volumeIsPublicRestricted :: Volume -> Bool
 volumeIsPublicRestricted v =
-  case volumePermissionPolicy v of
-    (PermissionPUBLIC, PublicRestricted) -> True
+  case volumeRolePolicy v of
+    RolePublicViewer PublicRestrictedPolicy -> True
+    RoleSharedViewer SharedRestrictedPolicy -> True
```
Comment (kanishka-azimi) [inline src/Databrary/Controller/Volume.hs]: Second real change.

File: src/Databrary/Model/Permission.hs [hunk 4]
```diff
@@ -49,74 +47,23 @@ releasePermission effectiveReleaseOnData currentUserAllowedPermissionOnVolume
...
-  dataPermission3 getObjEffectiveRelease getCurrentUserPermLevel obj > PermissionNONE
-
+    case getCurrentUserVolumeRole obj of
+      RolePublicViewer PublicRestrictedPolicy ->
+        releasePermission (effRelPrivate effRelease) PermissionPUBLIC
+      RoleSharedViewer SharedRestrictedPolicy ->
+        releasePermission (effRelPrivate effRelease) PermissionSHARED
+      -- other levels that behave more like private (options: none, shared, read, edit, admin) ?
+      rp ->
+         releasePermission (effRelPublic effRelease) (extractPermissionIgnorePolicy rp)
```
Comment (kanishka-azimi) [inline src/Databrary/Model/Permission.hs]: real change

**Comment by kanishka-azimi** [comment]
Aim to merge Friday or next week.

File: src/Databrary/Model/Permission.hs [hunk 5]
```diff
@@ -49,74 +47,23 @@ releasePermission effectiveReleaseOnData currentUserAllowedPermissionOnVolume
...
-      (p@PermissionPUBLIC, PublicRestricted) -> releasePermission (effRelPrivate effRelease) p
-      -- other levels that behave more like private (options: none, shared, read, edit, admin) ? 
-      (p, _) -> releasePermission (effRelPublic effRelease) p
-
-canReadData :: (a -> EffectiveRelease) -> (a -> (Permission, VolumeAccessPolicy)) -> a -> Bool
-canReadData getObjEffectiveRelease getCurrentUserPermLevel obj =
-  dataPermission3 getObjEffectiveRelease getCurrentUserPermLevel obj > PermissionNONE
-
+    case getCurrentUserVolumeRole obj of
+      RolePublicViewer PublicRestrictedPolicy ->
```
Comment (chreekat) [inline src/Databrary/Model/Permission.hs]: The names of these values causes some confusion.

Should I read line 56 as, "The current users's role wrt this volume is 'public viewer', and the policy affecting the user is 'public/restricted'"?

What does "PublicRestricted" even mean? It seems like an oxymoron.

`Role` is in the front, but `Policy` is in the back of the names?

File: src/Databrary/Model/Permission/Types.hs [hunk 6]
```diff
@@ -111,7 +114,29 @@ instance Monoid Access where
 
 deriveLiftMany [''Permission, ''Access]
 
-data VolumeAccessPolicy = PublicRestricted | PermLevelDefault
+data PublicPolicy = PublicRestrictedPolicy | PublicNoPolicy deriving (Show, Eq)
```
Comment (chreekat) [inline src/Databrary/Model/Permission/Types.hs]: Need Haddocks for new exports throughout this module.

File: src/Databrary/Model/Permission/Types.hs [hunk 7]
```diff
@@ -111,7 +114,29 @@ instance Monoid Access where
...
+
+data SharedPolicy = SharedRestrictedPolicy | SharedNoPolicy deriving (Show, Eq)
+
+data VolumeRolePolicy =
+    RoleNone
+  | RolePublicViewer PublicPolicy
+  | RoleSharedViewer SharedPolicy
+  | RoleReader
+  | RoleEditor
+  | RoleAdmin
```
[the remark this answers is not in the document]
Comment (chreekat) [reply src/Databrary/Model/Permission/Types.hs]: Seems nice

[the remark this answers is not in the document]
Comment (chreekat) [reply src/Databrary/Model/Permission/Types.hs]: I like the introduction of the word 'role'. I think that will clarify things a lot! Right now I still have trouble understanding all of the uses because I don't understand the "Policy" types.

File: src/Databrary/Model/Volume/Types.hs [hunk 8]
```diff
@@ -39,40 +38,38 @@ data Volume = Volume
...
+      RoleNone
     PermissionPUBLIC ->
       let shareFull = fromMaybe True mShareFull -- assume true because historically volumes were public full
-      in if shareFull then PermLevelDefault else PublicRestricted
-    _ ->
-      PermLevelDefault
-
+      in RolePublicViewer (if shareFull then PublicNoPolicy else PublicRestrictedPolicy)
+    PermissionSHARED ->
+      let shareFull = fromMaybe True mShareFull -- assume true because historically volumes were public full
```
Comment (chreekat) [inline src/Databrary/Model/Volume/Types.hs]: duplicate -- should be moved up

File: src/Databrary/Model/VolumeAccess/Types.hs [hunk 9]
```diff
@@ -26,8 +26,8 @@ getShareFullDefault targetParty individualAccessLevel =
     then nobodyPublicLegacyDefault
     else generalDefault
   where
-    getPartyId :: Party -> Id Party
-    getPartyId = partyId . partyRow
+    -- getPartyId :: Party -> Id Party
+    -- getPartyId = partyId . partyRow
```
Comment (chreekat) [inline src/Databrary/Model/VolumeAccess/Types.hs]: cleanup

File: src/Databrary/Model/VolumeState.hs [hunk 10]
```diff
@@ -24,7 +24,7 @@ lookupVolumeState :: (MonadDB c m) => Volume -> m [VolumeState]
 lookupVolumeState v = do
   let _tenv_a7xjl = unknownPGTypeEnv
   rows <- 
-    -- (selectQuery selectVolumeState "$WHERE volume = ${volumeId $ volumeRow v} AND (public OR ${volumePermission v >= PermissionEDIT})")
+    -- (selectQuery selectVolumeState "$WHERE volume = ${volumeId $ volumeRow v} AND (public OR ${(extractPermissionIgnorePolicy . volumeRolePolicy) v >= PermissionEDIT})")
```
Comment (chreekat) [inline src/Databrary/Model/VolumeState.hs]: cleanup

File: test/Databrary/Model/PermissionTest.hs [hunk 11]
```diff
@@ -21,3 +22,55 @@ unit_accessMember = do
     -- accessMember (TestContext { ctxIdentity = mkSession ... }) @?= PermissionNONE
     -- edge case
     accessMember (TestContext { ctxIdentity = IdentityNotNeeded }) @?= PermissionNONE
+
+runDataPermission4 :: (Release, Release) -> VolumeRolePolicy -> Permission
+runDataPermission4 (relPub, relPriv) rolePolicy =
+  dataPermission4 (const (EffectiveRelease {effRelPublic = relPub, effRelPrivate = relPriv})) (const rolePolicy) ()
+
+-- <volume shared level>AS<current user's access to that volume>
+fullySharedAsPublic, partiallySharedAsPublic, fullyOrPartiallyOrDatabrarySharedAsDatabraryMember :: VolumeRolePolicy
```
Comment (chreekat) [inline test/Databrary/Model/PermissionTest.hs]: With the right names for the Policy and Role constructors, these declarations would be superfluous.

File: test/Databrary/Model/PermissionTest.hs [hunk 12]
```diff
@@ -21,3 +22,55 @@ unit_accessMember = do
...
+publiclyReleasedExcerpt = (ReleasePUBLIC, ReleasePUBLIC)
+
+publiclyReleasedAsset, databraryReleasedAssetOrExcerpt, excerptsReleasedAssetOrExcerpt, privatelyReleasedAssetOrExcerpt :: (Release, Release)
+publiclyReleasedAsset = (ReleasePUBLIC, ReleasePRIVATE)
+excerptsReleasedAssetOrExcerpt = (ReleaseEXCERPTS, ReleasePRIVATE)  -- is it possible to set an excerpt to "excerpts' release level?
+databraryReleasedAssetOrExcerpt = (ReleaseSHARED, ReleasePRIVATE)  -- is this shared or private?
+privatelyReleasedAssetOrExcerpt = (ReleasePRIVATE, ReleasePRIVATE)
+
+-- TODO: use more idiomatic hunit
+unit_dataPermission4_combinations :: Assertion
```
Comment (chreekat) [inline test/Databrary/Model/PermissionTest.hs]: What sort of test is this? A figuring-things-out test? It looks like a reimplementation rather than a test.

Comment (kanishka-azimi) [reply test/Databrary/Model/PermissionTest.hs]: I'm not sure what to do with these any more.. I want to delete them, but I am looking for something to exercise similar logic. I may write some higher level BDD tests in ModelTest.hs to cover these cases.

I wrote these brief checks before we even had a unit test suite, to enumerate all the combinations I would like to have tested somehow. "dataPermission4" is one of the innermost points where a permission finally evaluated.

File: src/Databrary/Model/VolumeState.hs (hunk 10 above)
Comment (kanishka-azimi) [reply src/Databrary/Model/VolumeState.hs]: I've been leaving the old query in most cases, so it is easy to remember when two queries used to use the same "ORM" helper, such as `selectVolumeState` above. Once we have reached BEAM or postgresql-simple with some helpers, we can delete these.

File: src/Databrary/Model/Permission.hs (hunk 5 above)
Comment (kanishka-azimi) [reply src/Databrary/Model/Permission.hs]: Important, I'll try to get this to a point where it feels obvious.

File: test/Databrary/Model/PermissionTest.hs (hunk 12 above)
Comment (kanishka-azimi) [reply test/Databrary/Model/PermissionTest.hs]: Let's deal with this later. I spent some time on the documentation request for now.

File: src/Databrary/Model/Permission.hs (hunk 5 above)
Comment (kanishka-azimi) [reply src/Databrary/Model/Permission.hs]: Let's rename this next time this code change, or in at any point later when the code is at rest.

**Comment by kanishka-azimi** [comment]
I'll test the data migration and perform some light testing. Let me know if you want any more improvements or if any of the feedback you mentioned that I didn't address are critical to prevent merge.

File: test/Databrary/ModelTest.hs [hunk 13]
```diff
@@ -188,7 +188,7 @@ test_10 = Test.stepsWithTransaction "" $ \step cn2 -> do
     step "Then the lab B AI can't add volume acccess"
     -- Implementation of getVolume as used by postVolumeAccess
     Just volForAI2 <- runReaderT (lookupVolume ((volumeId . volumeRow) createdVol)) aiCtxt2
-    volumePermission volForAI2 @?= PermissionSHARED
```
Comment (kanishka-azimi) [inline test/Databrary/ModelTest.hs]: test7 onwards are meant to be the first-class citizen tests that deal with partially shared volumes as well as volumes shared other levels. I spent some time last week cleaning these up with the intention of adding a few more that focused on partially shared volumes, but I ran out of time. @chreekat

File: src/Databrary/Model/Volume/Types.hs [hunk 14]
```diff
@@ -53,25 +54,28 @@ instance Has Permission Volume where
...
+-- | Convert shareFull value read from db into a policy
+-- value, applying a default if needed.
+toPolicyDefaulting :: Maybe Bool -> a -> a -> a
+toPolicyDefaulting mShareFull noPolicy restrictedPolicy =
+    let
+        -- in the rare circumstance that a volume access
+        -- entry in db improperly contains null for public/shared group,
+        -- arbitrarily use True to follow old convention before sharefull
+        -- was introduced.
+        shareFull = fromMaybe True mShareFull
```
Comment (chreekat) [inline src/Databrary/Model/Volume/Types.hs]: We could migrate to a new, `not null` field...

File: src/Databrary/Model/Permission.hs [hunk 15]
```diff
@@ -6,11 +6,9 @@ module Databrary.Model.Permission
   , permissionPRIVATE
   , readPermission
   , readRelease
-  , dataPermission3
-  , canReadData
+  , dataPermission4
+  , canReadData2
   , accessJSON
-  -- testing only
```
[the remark this answers is not in the document]
Comment (chreekat) [reply src/Databrary/Model/Permission.hs]: ^ Quick win

File: src/Databrary/Model/Volume/Types.hs (hunk 14 above)
Comment (kanishka-azimi) [reply src/Databrary/Model/Volume/Types.hs]: I've been close to adding a default true not null... null is the best way to express how it is unspecified for most target parties currently, but I am hoping to finish the design by extending "share full" to apply to all volume access entries and switch to null without a default then.

File: schema/20180523-access-sharefull-contd.sql [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+-- migrate all existing entries such that share_full = true for db community group (0)
+--  when there is no entry indicating sharing is restricted for anonymous/nobody group (-1)
+update volume_access
+set share_full = true
+where (party, individual, children) = (0, 'SHARED', 'SHARED')
+and not exists
+  (select *
+   from volume_access va
+   where (va.volume, va.party, va.individual, va.children, va.share_full)
+     = (volume, -1, 'PUBLIC', 'PUBLIC', false));
```
Comment (kanishka-azimi) [inline schema/20180523-access-sharefull-contd.sql]: might not be referencing right volume column
